src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs

```diff
@@ -339,7 +340,7 @@ private void OnOriginFormTarget(TargetOffsetPathLength targetPath, Span<byte> ta
         var previousValue = _parsedRawTarget;
         if (ServerOptions.DisableStringReuse ||
             previousValue == null || previousValue.Length != target.Length ||
-            !StringUtilities.BytesOrdinalEqualsStringAndAscii(previousValue, target))
```

Comment:
This method does a null check as well, we'd need to make sure that's ok for all the changed callsites.

Reply:
> This method does a null check as well, we'd need to make sure that's ok for all the changed callsites.

I am not sure if I understand. So far all the callers of `StringUtilities.BytesOrdinalEqualsStringAndAscii` ensured that the input is not null. But if `null` would be sent to `StringUtilities.BytesOrdinalEqualsStringAndAscii` , it would throw. With my changes, it won't.

Would you like me to keep the old helper method that simply performs a debug assert for the input and calls the new Ascii API to do the job?

```cs
bool BytesOrdinalEqualsStringAndAscii(string previousValue, ReadOnlySpan<byte> newValue)
{
    Debug.Assert(previousValue is not null);

    return Ascii.Equals(previousValue, newValue);
}
```

Reply:
Sorry, I mistyped. I meant it checks for `0` not null.

Reply:
It is also done in BytesOrdinalEqualsStringAndAscii which is why I'm bringing it up
https://github.com/dotnet/aspnetcore/blob/main/src/Shared/ServerInfrastructure/StringUtilities.cs#L522

Reply:
I took a closer look at `BytesOrdinalEqualsStringAndAscii`.

If the remainder of the vectorized loop contains zero (or the input is simply too small to go the vectorized code path), but the inputs are equal it can still return `true`:

https://github.com/dotnet/aspnetcore/blob/39564d529f84f7a3bbac5b28ba11060e8ac30375/src/Shared/ServerInfrastructure/StringUtilities.cs#L501-L512

But the most important thing is that it checks for zero only one of the inputs:

https://github.com/dotnet/aspnetcore/blob/39564d529f84f7a3bbac5b28ba11060e8ac30375/src/Shared/ServerInfrastructure/StringUtilities.cs#L521-L522

And from what I can see the other input is always a const (known header for example):

https://github.com/dotnet/aspnetcore/blob/4e17e96258aebedd51d4dedb8fab201f7f3d600c/src/Servers/Kestrel/shared/KnownHeaders.cs#L326

It seems that only one of the inputs may contain null characters (because the other one is typically a pre-defined const) and hence `Ascii.Equals` will always return `false` in such cases because the inputs will simply not be equal?

@BrennanConroy is my understanding correct?